src/controller/CHIPDeviceController.cpp

```diff
@@ -1178,6 +1184,298 @@ void DeviceCommissioner::OnSessionEstablished()
+{
+    ChipLogProgress(Controller, "Device failed to receive the Attestation Information Response: 0x%02x", status);
+    DeviceCommissioner * commissioner = reinterpret_cast<DeviceCommissioner *>(context);
+    commissioner->mAttestationResponseCallback.Cancel();
+    commissioner->mOnAttestationFailureCallback.Cancel();
+    // TODO: Map error status to correct error code
+    commissioner->OnSessionEstablishmentError(CHIP_ERROR_INTERNAL);
+}
+
+void DeviceCommissioner::OnAttestationResponse(void * context, chip::ByteSpan attestationElements, chip::ByteSpan signature)
+{
+    ChipLogProgress(Controller, "Received Attestation Information from the device");
+    DeviceCommissioner * commissioner = reinterpret_cast<DeviceCommissioner *>(context);
+
+    commissioner->mAttestationResponseCallback.Cancel();
+    commissioner->mOnAttestationFailureCallback.Cancel();
+
+    if (commissioner->ValidateAttestationInfo(attestationElements, signature) != CHIP_NO_ERROR)
+    {
+        // Handle error, and notify session failure to the commissioner application.
+        ChipLogError(Controller, "Failed to validate the Attestation Information");
+        // TODO: Map error status to correct error code
+        commissioner->OnSessionEstablishmentError(CHIP_ERROR_INTERNAL);
+    }
+}
+
+CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(chip::ByteSpan attestationElements, chip::ByteSpan signature)
+{
+    VerifyOrReturnError(mState == State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mDeviceBeingPaired < kNumMaxActiveDevices, CHIP_ERROR_INCORRECT_STATE);
+
+    Device * device = &mActiveDevices[mDeviceBeingPaired];
+
+    ByteSpan certDeclaration;
+    ByteSpan attestationNonce;
+    ByteSpan timestamp;
+    ByteSpan firmwareInfo;
+    ByteSpan vendorReserved1;
+    ByteSpan vendorReserved2;
+    ByteSpan vendorReserved3;
+    ByteSpan vendorReserved4;
+
+    ByteSpan * element_array[] = { &certDeclaration, &attestationNonce, &timestamp,       &firmwareInfo,
+                                   &vendorReserved1, &vendorReserved2,  &vendorReserved3, &vendorReserved4 };
+
+    System::PacketBufferTLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+    System::PacketBufferHandle attestationElementsTbs;
+
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+    int i                      = 0;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+    do
+    {
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
+        const uint8_t * data = nullptr;
+        ReturnErrorOnFailure(tlvReader.GetDataPtr(data));
+        *element_array[i++] = ByteSpan(data, tlvReader.GetLength());
+    } while (tlvReader.Next() == CHIP_NO_ERROR);
```

Comment:
This will overrun `elements_array` if there are garbage context tags beyond the maximum. The loop needs to actively bound check and/or have necessary metadata as suggested above in `element_array` to detect seeing more tags than expected.

Reply:
updated to take this into account